README.rst
pykelihood
==========


.. image:: https://badge.fury.io/py/pykelihood.svg
    :target: https://pypi.org/project/pykelihood/

.. image:: https://img.shields.io/badge/pre--commit-enabled-brightgreen?logo=pre-commit&logoColor=white
   :target: https://github.com/pre-commit/pre-commit
   :alt: pre-commit

.. image:: https://img.shields.io/badge/code%20style-black-000000.svg
    :target: https://github.com/psf/black

-----
About
-----

Pykelihood is a Python package for statistical analysis designed to give more flexibility to likelihood-based inference
than is possible with **scipy.stats**. Distributions are designed from an Object Oriented Programming (OOP) point of
view.

Main features include:

- use any **scipy.stats** distribution, or make your own,
- fit distributions of arbitrary complexity to your data,
- add trends of different forms in the parameters of any distribution,
- condition the log-likelihood with any form of penalty,
- profile parameters with a penalised log-likelihood,
- more to come...


------------
Installation
------------

Using pip
------------------

.. code::

    pip install pykelihood


From sources
------------

.. code::

    git clone https://www.github.com/OpheliaMiralles/pykelihood

or

.. code::

    gh repo clone OpheliaMiralles/pykelihood


-----
Usage
-----

Basics
------

The most basic thing you can use ``pykelihood`` for is creating and manipulating distributions as objects.

>>> from pykelihood.distributions import Normal
>>> n = Normal(1, 2)
>>> n
Normal(loc=1.0, scale=2.0)

``n`` is an *object* of type ``Normal``. It has 2 parameters, ``loc`` and ``scale``. They can be accessed like standard
Python attributes:

>>> n.loc
1.0

Using the ``Normal`` object, you can calculate standard values using the same semantics as **scipy.stats**:

>>> n.pdf([0, 1, 2])
array([0.17603266, 0.19947114, 0.17603266])
>>> n.cdf([0, 1, 2])
array([0.30853754, 0.5       , 0.69146246])

Or you can also generate random values according to this distribution:

>>> n.rvs(10)
array([ 3.31370986,  5.02699468, -0.3573229 ,  1.00460378, -3.26044871,
        1.86362711, -0.84192901,  0.81132182, -2.03266978,  1.48079944])


Fitting
-------

Let's generate a larger sample from our previous object:

>>> data = n.rvs(1000)
>>> data.mean()
1.025039359276458
>>> data.std()
1.9376460645596842

We can fit a ``Normal`` distribution to this data, which will return another ``Normal`` object:

>>> Normal.fit(data)
Normal(loc=1.0250822420920338, scale=1.9376400770300832)

As you can see, the values are slightly different from the moments in the data.
This is due to the fact that the ``fit`` method returns the Maximum Likelihood Estimator (MLE)
for the data, and is thus the result of an optimisation (using **scipy.optimize**).

We can also fix the value for some parameters if we know them:

>>> Normal.fit(data, loc=1)
Normal(loc=1.0, scale=1.9377929687500024)

Trend fitting
*************

One of the most powerful features of ``pykelihood`` is the ability to fit arbitrary distributions.
For instance, say our data has a linear trend with a very little gaussian noise we would like to capture:

>>> import numpy as np
>>> data = np.linspace(-1, 1, 365) + np.random.normal(0, 0.001, 365)
>>> data[:10]
array([-0.99802364, -0.99503679, -0.98900434, -0.98277981, -0.979487  ,
       -0.97393519, -0.96853445, -0.96149152, -0.95564004, -0.95054887])

If we try to fit this without a trend, the resulting distribution will miss out on most of the information:

>>> Normal.fit(data)
Normal(loc=-3.6462053656578005e-05, scale=0.5789668679237372)

Let's fit a ``Normal`` distribution with a trend in the loc parameter:

>>> from pykelihood import kernels
>>> Normal.fit(data, loc=kernels.linear(np.arange(365)))
Normal(loc=linear(a=-1.0000458359290572, b=0.005494714384381866), scale=0.0010055323717468906)

``kernels.linear(X)`` builds a linear model in the form *a + bX* where *a* and *b* are parameters to
be optimised for, and *X* is some covariate used to fit the data. If we assume the data were daily observations,
then we find all the values we expected: *-1* was the value on the first day, *0.05* was the daily increment
(*2 / 365 = 0.05*), and there was a noise with std deviation *0.001*.


Fitting with penalties
**********************

Another useful feature of ``pykelihood`` is the ability to customize the log-likelihood function with penalties, conditioning methods, stability conditions, etc. Most statistics-related packages offer to fit data using the standard opposite log-likelihood function, or in the best case, preselected models. To our knowledge, ``pykelihood`` is the only Python package allowing to easily customize the log-likelihood function to fit data.

>>> data = np.random.normal(0, 1, 1000)
>>> def lassolike_score(distribution, data):
...     return -np.sum(distribution.logpdf(data)) + 5 * np.abs(distribution.loc())
...
>>> std_fit = Normal.fit(data)
>>> cond_fit = Normal.fit(data, score=lassolike_score)
>>> std_fit.loc.value
-0.010891307380632494
>>> cond_fit.loc.value
-0.006210406541824357

Parameter profiling
*******************

Likelihood based inference relies on parameter estimation. This is why it's important to quantify the sensitivity of a
chosen model to each of those parameters. The ``stats_utils`` module in ``pykelihood`` includes the ``Profiler``
class that allows to link a model to a set of observations by providing goodness of fit metrics and profiles for all
parameters.

>>> from pykelihood.profiler import Profiler
>>> from pykelihood.distributions import GEV
>>> fitted_gev = GEV.fit(data, loc=kernels.linear(np.linspace(-1, 1, len(data))))
>>> ll = Profiler(fitted_gev, data, inference_confidence=0.99) # level of confidence for tests
>>> ll.AIC  # the standard fit is without trend
{'AIC MLE': -359.73533182968777, 'AIC Standard MLE Fit': 623.9896838880583}
>>> ll.profiles.keys()
dict_keys(['loc_a', 'loc_b', 'scale', 'shape'])
>>> ll.profiles["shape"].head(5)
      loc_a     loc_b     scale     shape   likelihood
0 -0.000122  1.000812  0.002495 -0.866884  1815.022132
1 -0.000196  1.000795  0.001964 -0.662803  1882.043541
2 -0.000283  1.000477  0.001469 -0.458721  1954.283256
3 -0.000439  1.000012  0.000987 -0.254640  2009.740282
4 -0.000555  1.000016  0.000948 -0.050558  1992.812843

Confidence intervals can be computed for specified metrics:

>>> def metric(gev): return gev.loc()
>>> ll.confidence_interval(metric)
[-4.160287666875364, 4.7039931595123825]


------------
Contributing
------------

`Poetry <http://python-poetry.org>`_ is used to manage ``pykelihood``'s dependencies and build system. To install
Poetry, you can refer to the `installation instructions <https://python-poetry.org/docs/#installation>`_, but it boils
down to running:

.. code::

    curl -sSL https://raw.githubusercontent.com/python-poetry/poetry/master/get-poetry.py | python



To configure your environment to work on pykelihood, run:

.. code-block::

    git clone https://www.github.com/OpheliaMiralles/pykelihood  # or any other clone method
    cd pykelihood
    poetry install

This will create a virtual environment for the project and install the required dependencies. To activate the virtual
environment, be sure to run :code:`poetry shell` prior to executing any code.

We also use the `pre-commit <https://pre-commit.com>`_ library which adds git hooks to the repository. These must be installed with:

.. code::

   pre-commit install

Some parts of the code base use the `matplotlib <https://matplotlib.org/>`_ and
`hawkeslib <https://hawkeslib.readthedocs.io/en/latest/index.html>`_ package, but are for now not required to run most
of the code, including the tests.

Tests
-----

Tests are run using `pytest <https://docs.pytest.org/en/stable/>`_. To run all tests, navigate to the root folder or the
``tests`` folder and type :code:`pytest`.


CHANGELOG.md
# Changelog

## 0.4.1 - 2023-10-25

### New features

* The `Distribution.fit` method accepts a `scipy_args` dictionary which is
  passed to `scipy`'s `minimize` function.
* The confidence interval computed by the profiler now uses root finding to
  find the bounds where the likelihood ratio test starts failing. This means
  confidence intervals can only be computed for the distribution's parameters.
* Upper bounds on dependencies were removed, improving compatibility with
  recent versions.

## 0.4.0 - 2023-10-14

### Breaking changes

* Renamed `stats_utils` module to `profiler`
* Data must now be provided to kernels on creation, unbound kernels are
  no longer allowed
* `Parameter`s are no longer subclasses of `float`, use `.value` to get
  their stored value
* `ConditioningMethod`s were removed, their uses can be replaced with
  _score functions_
* The `biv` parameter to the `Profiler` was removed, confidence
  intervals are univariate only

### Removed

Many distributions and utilities which were created with a specific use
case in mind and aren't generally useful have been removed:

* `MixtureExponentialModel`,
* `ExtendedGPD`,
* `PointProcess`,
* `CompositionDistribution`,
* `DetrendedFluctuationAnalysis`,
* `pettitt_test`,
* `threshold_selection_GoF` and `threshold_selection_gpd_NorthorpColeman`,
* extreme values visualisation routines,
* process samplers (Poisson and Hawkes).

### New features

* Metrics: `{pp,qq}_l{1,2}_distance`, `likelihood`, `expo_ratio`
* Log-normal distribution
* Plotting functions now accept an `ax` argument to use instead of the
  global `plt` figure
* Constant kernel (most useful for testing)
* `Kernel`s have a `with_covariate` method that returns a new kernel
  with the provided data as covariate, but all parameters are kept the
  same
* The `random_state` parameter to the `Distribution.rvs` method is now
  explicit and no longer hidden in the `**kwargs`

### Bug fixes

* Fixed `fit_instance` for nested kernels with fixed values
* Fixed the `TruncatedDistribution` which forgot its bounds after fitting
* A parameter which shows up in several places in a distribution will
  keep the same value when fitting instead of returning independent
  parameters

### Other

* Add section to README on fitting other score functions than the likelihood
* Add changelog with all version changes up to this one

## 0.3.2 - 2021-03-26

Various bug fixes due to new names in version 0.3.0

## 0.3.1 - 2021-03-26

### New features

* New trigonometric kernel

## 0.3.0 - 2021-03-23

This release aims at making fitting even more generic by replacing
penalized likelihoods with score functions

### Breaking changes

* Remove `log_likelihood` and `opposite_log_likelihood` methods from
 `Distribution`s
* Remove `penalty` argument from `fit`
* `ConditioningMethod`s take `distribution` as the first argument and
 `data` as the second
* Rename `Likelihood` class to `Profiler`
* Rename `mle` attribute to `optimum` in `Profiler` class

### New features

* `fit*` methods minimize a `score` function which takes a distribution
  and data as arguments. By default, they maximize the log likelihood.
* New `metrics` module which contains scoring functions
* The `Profiler` has a `score_function` argument
* Add `threshold_selection` based on a multiple threshold penultimate model
* New QQ-plot visualization methods

### Bug fixes
* `Distribution.rvs` now passes all parameters to `scipy`'s `rvs` methods.
  In particular, `random_state` is now propagated.
* `TruncatedDistribution` now ignores data outside its range

### Other

* Improve overall typing of the library

## 0.2.1 - 2021-02-21

### New features

* Add `TruncatedDistribution`

### Bug fixes

* Avoid replacing non-constant parameters with `ConstantParameter`s
  inside `ParametrizedFunction`s
* Fix nested kernel fitting

## 0.2.0 - 2021-02-09

This release aims at making the fitting feature more generic by allowing
other forms of likelihood conditioning and other metrics than the return
level in the confidence interval computation.

### Breaking changes

* Remove `simulations` and `stopping_times` modules
* Removed `Likelihood.return_level`
* Move `visualisation_utils` to their own subpackage
* Rename `conditioning_method` argument to `penalty`
* Rename `profile_likelihood` method to `fit_instance`

### New features

* New method `Likelihood.confidence_interval` with the `metric` argument
  to control which value to estimate

### Other

* Add section to README.rst about parameter profiling
* Add pre-commit hooks
* Fix README.rst directives


## 0.1.0 - 2021-02-04

Initial release


pykelihood/profiler.py
from __future__ import annotations

import math
import warnings
from itertools import count
from typing import Callable, Tuple

import numpy as np
import pandas as pd
from scipy.optimize import brentq
from scipy.stats import chi2

from pykelihood.distributions import Distribution
from pykelihood.metrics import opposite_log_likelihood

try:
    from functools import cached_property
except ImportError:
    from cached_property import cached_property

warnings.filterwarnings("ignore")


class Profiler(object):
    def __init__(
        self,
        distribution: Distribution,
        data: pd.Series,
        score_function: Callable = opposite_log_likelihood,
        name: str = "Standard",
        inference_confidence: float = 0.99,
        single_profiling_param=None,
        optimization_method="Nelder-Mead",
        x0=None,
    ):
        """l

        :param distribution: distribution on which the inference is based
        :param data: variable of interest
        :param score_function: function used for optimisation
        :param name: name (optional) of the profile if it needs to be compared to other score functions
        :param inference_confidence: wanted confidence for intervals
        :param single_profiling_param: parameter that we want to fix to create the profiles based on the score function
        """
        self.name = name
        self.distribution = distribution
        self.data = data
        self.score_function = score_function
        self.inference_confidence = inference_confidence
        self.single_profiling_param = single_profiling_param
        self.optimization_method = optimization_method
        self.x0 = x0

    @cached_property
    def standard_mle(self):
        estimate = self.distribution.fit(self.data, method=self.optimization_method)
        ll = -opposite_log_likelihood(estimate, self.data)
        ll = ll if isinstance(ll, float) else ll[0]
        return (estimate, ll)

    @cached_property
    def optimum(self):
        estimate = self.distribution.fit_instance(
            self.data,
            score=self.score_function,
            x0=self.x0,
            scipy_args={"method": self.optimization_method},
        )
        func = -self.score_function(estimate, self.data)
        func = func if isinstance(func, float) else func[0]
        return (estimate, func)

    @cached_property
    def profiles(self):
        profiles = {}
        opt, func = self.optimum
        if self.single_profiling_param is not None:
            params = [self.single_profiling_param]
        else:
            params = [n[0] for (v, n) in opt.param_mapping()]
        for name, k in opt.optimisation_param_dict.items():
            if name in params:
                r = float(k)
                b = 2 * 10 ** (math.floor(math.log10(np.abs(r))) - 1)
                range = np.linspace(r - b, r + b, 40)
                profiles[name] = self.test_profile_likelihood(range, name)
        return profiles

    def test_profile_likelihood(self, range_for_param, param):
        opt, func = self.optimum
        profile_ll = []
        params = []
        for x in range_for_param:
            try:
                pl = opt.fit_instance(
                    self.data,
                    score=self.score_function,
                    **{param: x},
                )
                pl_value = -self.score_function(pl, self.data)
                pl_value = pl_value if isinstance(pl_value, float) else pl_value[0]
                if np.isfinite(pl_value):
                    profile_ll.append(pl_value)
                    params.append([p.value for p in pl.flattened_params])
            except:
                pass
        chi2_par = {"df": 1}
        lower_bound = func - chi2.ppf(self.inference_confidence, **chi2_par) / 2
        filtered_params = pd.DataFrame(
            [x + [eval] for x, eval in zip(params, profile_ll) if eval >= lower_bound]
        )
        cols = list(opt.flattened_param_dict.keys()) + ["score"]
        filtered_params = filtered_params.rename(columns=dict(zip(count(), cols)))
        return filtered_params

    def confidence_interval(self, param: str, precision=1e-5) -> Tuple[float, float]:
        opt, func = self.optimum
        value_threshold = func - chi2.ppf(self.inference_confidence, df=1) / 2

        def score(x: float):
            new_opt = opt.fit_instance(
                self.data, score=self.score_function, **{param: x}
            )
            return -self.score_function(new_opt, self.data)

        def delta_to_threshold(x: float):
            return score(x) - value_threshold

        def is_outside_conf_interval(x: float):
            return score(x) < value_threshold

        param_value = opt.flattened_param_dict[param].value
        if is_outside_conf_interval(param_value):
            raise RuntimeError("Optimum is outside its own confidence interval?")

        step = 0.1 * abs(param_value)
        lb_underestimate = param_value - step
        ub_overestimate = param_value + step
        for _ in range(1000):
            if is_outside_conf_interval(lb_underestimate):
                break
            lb_underestimate -= step
        else:  # nobreak
            raise RuntimeError("Unable to find confidence interval lower bound")

        for _ in range(1000):
            if is_outside_conf_interval(ub_overestimate):
                break
            ub_overestimate += step
        else:  # nobreak
            raise RuntimeError("Unable to find confidence interval upper bound")

        lb = brentq(delta_to_threshold, lb_underestimate, param_value, xtol=precision)
        ub = brentq(delta_to_threshold, param_value, ub_overestimate, xtol=precision)

        return lb, ub

    # For compatibility
    confidence_interval_bs = confidence_interval
